### PR TITLE
snap-confine/nvidia: Support legacy biarch trees for GLVND systems

### PR DESCRIPTION
**Pull request by ikeydoherty**
GLVND systems are well catered for on biarch configurations, however the
older 304 + 340 series will rely on ld.so.conf tricks to mask the system
libGL.so with one in the private `libdir/nvidia*/` files. This is seen
consistently across Arch, Fedora and Solus. As such we allow these fallback
paths to work so that users of snapd on glvnd enabled distros with the older
legacy drivers can still enjoy snapd + NVIDIA together.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>

### CODE REVIEW

**Comment by ikeydoherty** [comment]
Sample `/var/lib/snapd/lib/gl` on a GLVND distro (Solus) with the NVIDIA 304 driver: https://hastebin.com/idasofebut - should point out that the primary links are what matter here. Also note NVIDIA 340 provides the libEGL links which in turn would replace the glvnd libEGL in the hostfs portion. And "pure" GLVND still works.

**Comment by Conan-Kudo** [comment]
Oh goodie, this should fix some of the reported issues in Fedora with NVIDIA, too.

**Comment by Conan-Kudo** [comment]
@mvo5 Could you please cherry pick this into 2.31 for me? This will resolve [RH#1442051](https://bugzilla.redhat.com/show_bug.cgi?id=1442051) for me.

File: cmd/snap-confine/mount-support-nvidia.c [hunk 1]
```diff
@@ -216,6 +224,14 @@ static void sc_populate_libgl_with_hostfs_symlinks(const char *libgl_dir,
 				 "%s/%s", libgl_dir, filename);
 		debug("creating symbolic link %s -> %s", symlink_name,
 		      symlink_target);
+
+		// Make sure we don't have some link already (merged GLVND systems)
+		if (lstat(symlink_name, &stat_buf) == 0) {
+			if (unlink(symlink_name) != 0) {
```
Comment (jdstrand) [inline cmd/snap-confine/mount-support-nvidia.c]: Shouldn't you also be testing if S_IFLNK? lstat() will happily return 0 if regular file or symlink.

Comment (ikeydoherty) [reply cmd/snap-confine/mount-support-nvidia.c]: The path is inside our hostfs symlink farm that we create, and we only create symlinks, the code doesn't copy files.

Comment (ikeydoherty) [reply cmd/snap-confine/mount-support-nvidia.c]: Note, the naming here is confusing, but `symlink_name` is **not** the source name, it is the **target** name for the symlink, i.e. the new inode. Confused the hell out of me going through it the first time.

**Comment by Conan-Kudo** [comment]
@ikeydoherty I think I might be confusing myself again, but will this actually work with legacy nvidia driver stuff installed in `/usr/lib64/nvidia*`?

c.f.: http://koji.rpmfusion.org/koji/rpminfo?rpmID=171143
c.f.: http://koji.rpmfusion.org/koji/rpminfo?rpmID=213317

**Comment by ikeydoherty** [comment]
Uh it looks like you guys have even more stuff in that directory than I suspected. You have files in there that shouldn't and don't need to be.. Only the libGL needs to be there :/

In that case we'd have to effectively **double** the copy list to include all of those (frankly incredibly silly) paths.

**Comment by ikeydoherty** [comment]
And TBH @Conan-Kudo it looks like Fedora should be using bind mounts for those nvidia drivers, but my guess is the use of glvnd makes that unfeasible.

What we can do:

Remove 32 vs 64 logic in the "mount" farm. Make the function accept a common ancestor, i.e. `/usr/lib` `/usr/lib/nvidia`, `/usr/lib32`, etc. Then form full paths from that ancestor and do the blits, while preserving the above "nuke if exists" logic.

That should simplify the code *and* cover all of the uses, right? And keep the copy list simpler by implicitly allowing skipping over elements.

**Comment by ikeydoherty** [comment]
OK I added a follow up commit which should just make stuff work for Everyone ™

**Comment by ikeydoherty** [comment]
The only pain point I see (because I don't know how it works on Fedora) is this reliance on `/usr/lib` and not, say, `/usr/lib64`. Now, if this is an issue on Fedora (i.e they're distinct directories maybe?) then we can make it a configure time option for the LIBDIR? Or patch just patch that one spot, maybe.

**Comment by Conan-Kudo** [comment]
@ikeydoherty They are distinct directories. `/usr/lib` is 32-bit, whereas `/usr/lib64` is 64-bit.

We could make this a compile-time option if you want, then we can feed in the correct base libdir as a configure option (we already do, it just does nothing with it...).

File: cmd/snap-confine/mount-support-nvidia.c (hunk 1 above)
Comment (zyga) [reply cmd/snap-confine/mount-support-nvidia.c]: I'll probably do a small update here to use consistent terminology (oldname vs whatever the other one was).

**Comment by ikeydoherty** [comment]
Yay! Danke ^_^

**Pull request merged**